### ISSUE TITLE
systemd-logind takes over 1 s on current device

### ISSUE THREAD

**Issue by paulmenzel** [systemd/systemd#17881]
On a current Intel Tiger Lake Dell XPS 13 9310 2-in-1/02P55J, BIOS 1.0.3 10/05/2020 with Debian sid/unstable, according to `systemd-analyze`, systemd-logind takes over *one* second to start. (It look unrelated to button detection.)

    $ systemd --version
    systemd 247 (247.1-3)
    +PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid

```
$ systemd-analyze critical-chain 
The time when unit became active or started is printed after the "@" character.
The time the unit took to start is printed after the "+" character.

graphical.target @1.953s
└─multi-user.target @1.953s
  └─unattended-upgrades.service @1.953s
    └─systemd-logind.service @807ms +1.144s
      └─basic.target @789ms
        └─sockets.target @789ms
          └─dbus.socket @789ms
            └─sysinit.target @786ms
              └─systemd-backlight@leds:dell::kbd_backlight.service @1.731s +46ms
                └─system-systemd\x2dbacklight.slice @1.731s
                  └─system.slice @206ms
                    └─-.slice @206ms
$ journalctl -o short-precise -u systemd-logind -b
-- Journal begins at Thu 2020-12-03 16:37:46 CET, ends at Mon 2020-12-07 10:34:47 CET. --
Dec 07 09:07:06.477758 schuetz systemd[1]: Starting User Login Management...
Dec 07 09:07:07.617624 schuetz systemd-logind[1563]: New seat seat0.
Dec 07 09:07:07.619564 schuetz systemd-logind[1563]: Watching system buttons on /dev/input/event4 (Intel HID events)
Dec 07 09:07:07.621138 schuetz systemd-logind[1563]: Watching system buttons on /dev/input/event0 (AT Translated Set 2 keyboard)
Dec 07 09:07:07.621395 schuetz systemd-logind[1563]: Watching system buttons on /dev/input/event5 (Intel HID 5 button array)
Dec 07 09:07:07.621636 schuetz systemd-logind[1563]: Watching system buttons on /dev/input/event7 (Power Button)
Dec 07 09:07:07.621854 schuetz systemd-logind[1563]: Watching system buttons on /dev/input/event6 (Lid Switch)
Dec 07 09:07:07.622486 schuetz systemd[1]: Started User Login Management.
Dec 07 09:07:07.673273 schuetz systemd-logind[1563]: Watching system buttons on /dev/input/event8 (Sleep Button)
Dec 07 09:07:08.305531 schuetz systemd-logind[1563]: New session c1 of user Debian-gdm.
```

This is a user device, so it is hard to get timely debug information. I am trying to get the debug logs.

My suspicion is, that it has to do with the security features of the service unit taking a long time to set up.

https://github.com/systemd/systemd/blob/e0f90ad98802a43addd40f7f8ef69118d04f4cd2/units/systemd-logind.service.in#L26-L60

**Comment by bluca** [comment]
Any chance of some profiling data on top of debug logs?

**Comment by paulmenzel** [comment]
> Any chance of some profiling data on top of debug logs?

Please tell me how to collect profiling data. (But as written, it’ll be hard as it’s in use by somebody else.)

**Comment by bluca** [comment]
> > Any chance of some profiling data on top of debug logs?
> 
> Please tell me how to collect profiling data. (But as written, it’ll be hard as it’s in use by somebody else.)

Would need to run through perf or similar, so if it's on a user device it's going to be challenging at best. Let's wait for debug logs before going down that road, hopefully they'll be enough to pin-point where the slowdown happens.

**Comment by poettering** [comment]
I am sorry, but such vague bug reports and not helpful and not actionable. Please properly profile things, prepare patches and such. Just filing bugs about assumed slowness helps noone, and it is unlikely anyone of the maintainers will really care much, since 100ms that is run in parallel with a lot of other stuff really doesn't matter.

Our primary goal is correctness and well-definedness, sheer performance is just secondary to that, and will always take the backseat.

I mean, I welcome optimization work, if it doesn't compromise correctness. But I am not too keen on bug reports that aren't about actual bugs nor directly actionable.

**Comment by paulmenzel** [comment]
I understand, but as doing this only as a side activity, that is out of my time scope. The next problem is, that I often only have access at install time to the devices, and then they are given out.

What would help and make it easier, if profiling was integrated into systemd, so I only have to pass a Linux command line option or edit `/etc/systemd/system.conf` and the data would be shown, instead of fiddling around with perf, oprofile, and what not.

**Comment by bluca** [comment]
Have you looked at https://github.com/systemd/systemd-bootchart ?
It gives you a process-level view, it might not be granular enough for this case, but it could be a start.

**Comment by paulmenzel** [comment]
> Have you looked at https://github.com/systemd/systemd-bootchart?

Yes, I have. (I think, I am the main user of systemd-bootchart. ;-))

> It gives you a process-level view, it might not be granular enough for this case, but it could be a start.

If I haven’t missed an option of the tool, it’s not applicable to this purpose. It does not show what is happening in between:

    Dec 07 09:07:06.477758 schuetz systemd[1]: Starting User Login Management...
    Dec 07 09:07:07.617624 schuetz systemd-logind[1563]: New seat seat0.

I think, the debug output is going to help. I am going to try to retrieve it.

**Comment by bluca** [comment]
> > Have you looked at https://github.com/systemd/systemd-bootchart?
> 
> Yes, I have. (I think, I am the main user of systemd-bootchart. ;-))

Oh I know a few others :-D

> > It gives you a process-level view, it might not be granular enough for this case, but it could be a start.
> 
> If I haven’t missed an option of the tool, it’s not applicable to this purpose. It does not show what is happening in between:
> 
> ```
> Dec 07 09:07:06.477758 schuetz systemd[1]: Starting User Login Management...
> Dec 07 09:07:07.617624 schuetz systemd-logind[1563]: New seat seat0.
> ```
> 
> I think, the debug output is going to help. I am going to try to retrieve it.

You haven't missed anything, the granularity is per process. It would have been useful in case logind was stuck waiting for some dependency or such cases.
